sklearn/decomposition/__init__.py

```diff
@@ -4,19 +4,30 @@
 this module can be regarded as dimensionality reduction techniques.
 """
 
-from ._nmf import NMF, non_negative_factorization
-from ._pca import PCA
-from ._incremental_pca import IncrementalPCA
-from ._kernel_pca import KernelPCA
-from ._sparse_pca import SparsePCA, MiniBatchSparsePCA
-from ._truncated_svd import TruncatedSVD
-from ._fastica import FastICA, fastica
+# TODO: remove me in 0.24:
+# pre-cache the import of the deprecated module to avoid showing
+# the partial_dependence function.
+# https://github.com/scikit-learn/scikit-learn/issues/15842
+import warnings
+with warnings.catch_warnings():
+    warnings.simplefilter("ignore", category=FutureWarning)
+    from . import dict_learning as _ignored
+    del _ignored
```

Comment:
is there a way to also remove the module from the cache, so that warnings are still raised the next time it's imported? I think I know the answer but I'm still in the denial phase

Reply:
But it is not cached, it will be reported and then the `dict_learning` function will be hidden again by the module. I have tried playing with `__getattr__` at the `sklearn.decomposition` package level but I could not achieve what I wanted.

Reply:
Are you sure?

My understanding is that in 0.21, the reason `sklearn.decomposition.dict_learning` returns the function instead of the module is because the *function* is cached, not the module.

Reply:
But then the `sklearn/decomposition/__init__.py` file is never reexecuted because the `sklearn.decomposition` is imported only once. This is what I meant by cached.